### DISCUSSION TITLE
Concurrent mutations hang with await Promise.all

### DISCUSSION THREAD

**Discussion by 1timeSports** [urql-graphql/urql#2636] [Q&A] [answered]
Hello,

Wasn't sure if this is actually an issue or just something I'm missing, so decided to start a discussion.

I have noticed some odd behavior with react hook mutations.

My client looks like this:

```
const gqlClient = createClient({
  url,
  exchanges: [
    cacheExchange,
    authExchange,
    fetchExchange,
  ],
});
```

When I run this mutation 5 times with concurrently fired promises, only one request is actually sending.  At first I thought this might be the dedupExchange doing its thing, but I removed it from my urql client and still get the same behavior.

```
mutation GenerateImageUploadUrl {
    generateImageUploadUrl {
        id
        uploadURL
    }
}
```

I use codegen'd react hooks for this so I end up with some code that looks like this:

```
const awaitTimeout = () =>
  new Promise((resolve) => {
    setTimeout(() => resolve(true), 1000);
  });

const [, generateImageUploadUrl] = useGenerateImageUploadUrl();

const handleCreateTest = useCallback(async () => {
  // HANGS, NO RESOLUTION OR REJECTION
  const promises = [1, 2, 3, 4, 5].map(() => generateImageUploadUrl({}));

  // SUCCEEDS
  // const promises = [1, 2, 3, 4, 5].map(() => awaitTimeout());

  console.log('starting promises');

  try {
    await Promise.all(promises);
  } catch (err) {
    console.log('err', err);
  }

  console.log('done with promises');
}, [generateImageUploadUrl]);
```

Dependencies:
```
    "urql": "3.0.1",
    "@urql/core": "^3.0.1",
    "@urql/devtools": "^2.0.3",
    "@urql/exchange-auth": "^1.0.0",
    "@urql/exchange-graphcache": "5.0.0"
```

**Comment by kitten** [comment, marked as answer]
Didn't take a full look yet, but at first glance, you're trying to use the `useMutation` hook, which is rather built to deliver one result rather than `client.mutation(...).to promise()`, which are the "direct" calls which are easier to use if you need multiple results 

As per the hook, it shouldn't stall though, so I'll check what's going on with that later

Edit: Also, btw, can you please supply some info on which version of `urql` and `@urql/core` you're using please?

**Reply by austin43** [reply]
@kitten OP here. Thanks for the quick response. 

Dep versions are in the bottom of original post. 

I will give client.mutation a try in a bit.

**Reply by austin43** [reply]
@kitten I gave client.mutation a shot, and I still get the issue of some requests not going out randomly.  It is better, as in I get 1/10 failures instead of 8/10, but some requests still aren't going out.

**Reply by kitten** [reply]
At this point, I can't really help you much more 😅 I wrote a quick unit test and everything's working correctly. In fact, we enabled concurrent, identical mutations a couple months ago. So, if you can isolate this in a reproduction, I can take a closer look.

**Reply by austin43** [reply]
@kitten That's completely fair.  Good to know concurrent identical mutations are valid now.  I'll try and get a valid repro case, although I'm guessing it's just something about my setup that is causing this.  Thank you for looking into this 😄
